### ISSUE TITLE
Non-iid noise in accelerated training code

### ISSUE THREAD

**Issue by caelum02** [VsonicV/es-fine-tuning-paper#46]
From utils/worker_extn.py:L23 
```
def perturb_self_weights(self, seed, noise_scale, negate=False):
    scale = float(noise_scale)
    sign = -1.0 if negate else 1.0
    for _, p in self.model_runner.model.named_parameters():
        gen = torch.Generator(device=p.device)
        gen.manual_seed(int(seed))
        noise = torch.randn(p.shape, dtype=p.dtype, device=p.device, generator=gen)
        p.data.add_(sign * scale * noise)
        del noise
    if torch.cuda.is_available():
        torch.cuda.synchronize()
    torch.cuda.empty_cache()
    return True
```

Here the same seed is shared between different layers, resulting in non-independent perturbations

Same issue persists in non-accelerated countdown training code

```
    # Weight Perturbation
    for name, param in model.named_parameters():
        gen = torch.Generator(device=param.device)

        gen.manual_seed(int(seed))

        noise = torch.randn(
            param.shape,
            generator=gen,
            device=param.device,
            dtype=param.dtype
        )
        param.data.add_(SIGMA * noise)
```

I guess this was an intended behavior since there exists iid version training code in the repository - countdown/es_fine-tuning_countdown_iid.py, but isn't this a bug? Any effects on train results - as in theory the noise should be i.i.d.

p.s. checked #3 but seems like fixing this issues wouldn't result in perturbation & restoration inconsistency

**Comment by VsonicV** [comment]
@caelum02 This is not a bug, and we have clarified this point in our paper as well (the "Partially correlated noise vs. i.i.d. noise." paragraph in Appendix A.2). For our testing, we did not observe a significant performance difference between non-iid and iid implementations, though the iid implementation seems to have a slightly higher variance across runs. The non-iid version gave us some convenience in implementation and a potential speed-up via caching the repeated noise, though it does not strictly match the theoretical assumption. Empirically, these two versions do not differ too much.

For #3 , the first proposed fix in that PR is indeed buggy and can cause perturbation & restoration inconsistency, so we implemented a correct iid version by ourselves and added it to the main repo.
